### PR TITLE
AP audience (visibility) パースの修正

### LINKED ISSUE #4951: Accept all three representations of the Public collection
## Summary

https://github.com/syuilo/misskey/blob/9e20fc5c88683d92a0bd9aade4ef6085bfb27e8c/src/remote/activitypub/models/note.ts#L117-L118

> ##### NOTE
> Compacting an ActivityStreams object using the ActivityStreams JSON-LD context might result in https://www.w3.org/ns/activitystreams#Public being represented as simply Public or as:Public which are valid representations of the Public collection. Implementations which treat ActivityStreams objects as simply JSON rather than converting an incoming activity over to a local context using JSON-LD tooling should be aware of this and should be prepared to accept all three representations. 
&mdash; [5.6 Public Addressing - ActivityPub](https://www.w3.org/TR/activitypub/#x5-6-public-addressing)

### LINKED ISSUE #5078: Support "Create=>Image" activity
## Summary
Misskey should display the activity sent with ActivityPub when someone create an image.
See https://github.com/tootsuite/mastodon/issues/11137

### LINKED ISSUE #5767: hubzilla public posts are shown as private notes
hubzilla is another fediverse network which additionally supports activitypub.  [Here](https://yuzulia.xyz/notes/816fd9495f30b4301cb8f1e7) the original  post is public but not visible. Mastodon side people can see it.

is this hubzilla's problem?

### LINKED ISSUE #5768: audienceのないAP Objectの公開範囲の扱い
## Summary
AP Note等に audience (to, cc) がなくて、Private扱いになってしまう投稿等がある、どうするか。

1. HubzillaにAP Noteを取得しに行った場合に
Note オブジェクトにaudience(to, cc)が明示されていないため、Privateとして扱われてしまう。

2. Hubzillaから、Create ActivityでImageが来た場合に
Image オブジェクトにaudience(to, cc)が明示されていないため、Privateとして扱われてしまう。
※ 確認は出来ないが Create => Note の場合は、ちゃんと来ているので大丈夫な気がする。

Related
https://github.com/syuilo/misskey/issues/5767
https://github.com/tootsuite/mastodon/issues/11137

### PR DESCRIPTION
**Pull request by mei23**
## Summary
Fix #4951
Resolve #5768
Resolve #5078
Resolve #5767

- audience (visibility) あたりのリファクタ (#4951のPublic表現の修正含む)
- Create => Note, Image ... 等で、Note等のObjectにaudience (to, cc) がない場合は、
  Activity の audience を見に行くように
- Reply, Renote 等でNote等をremoteに取得しに行った場合に audience がない場合は、
  匿名で取得できたのならば公開されているものに違いないとみなして public 扱いにするように。

### CODE REVIEW

File: src/remote/activitypub/audience.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+type Visibility = 'public' | 'home' | 'followers' | 'specified';
+
+type AudienceInfo = {
+	visibility: Visibility,
+	mentionedUsers: User[],
+	visibleUsers: User[],
+};
+
+export async function parseAudience(actor: IRemoteUser, to: ApObject | undefined, cc: ApObject | undefined, resolver?: Resolver): Promise<AudienceInfo> {
```
Comment (acid-chicken) [inline src/remote/activitypub/audience.ts]:
```suggestion
export async function parseAudience(actor: IRemoteUser, to?: ApObject, cc?: ApObject, resolver?: Resolver): Promise<AudienceInfo> {
```

もし省略不能を意識させるのを意図した設計なら `null` の方がいいかもと思ったけど `getApIds` も触る必要が出てくるので、もし前述の通りであればとりあえず無視して Resolve してください。

File: src/remote/activitypub/models/note.ts [hunk 2]
```diff
@@ -109,25 +110,24 @@ export async function createNote(value: string | IObject, resolver?: Resolver, s
...
+	// Audience (to, cc) が指定されてなかった場合
+	if (visibility === 'specified' && visibleUsers.length === 0) {
+		if (activity && isCreate(activity)) {
+			// Create 起因ならば Activity を見る
+			const activityAudience = await parseAudience(actor, activity.to, activity.cc);
+			visibility = activityAudience.visibility;
+			visibleUsers = activityAudience.visibleUsers;
+			apMentions = activityAudience.mentionedUsers;
+		} else if (typeof value === 'string') {	// 入力がstringならばresolverでGETが発生している
+			// こちらから匿名GET出来たものならばpublic
```
[the remark this answers is not in the document]
Comment (mei23) [reply src/remote/activitypub/models/note.ts:127]: 公開範囲=未定義 ってことなのでなんともいえない

最初homeが無難かなとも思ったのですが
元々公開するための仕組みで公開範囲の概念がないのがデフォルトなのでpublicでいいかなと思いました。

あとhomeにすると、それを参照するRenote/Reply自体の公開範囲が常にhomeになっちゃいます。

File: src/remote/activitypub/audience.ts (hunk 1 above)
Comment (mei23) [reply src/remote/activitypub/audience.ts]: `なんたら | undefined` は`type`だとlintに指摘されるのに`function`だと指摘してくれない…

**Review by syuilo — APPROVED** [review]

**Comment by syuilo** [comment]
ありがとうございます 🙏

File: src/remote/activitypub/audience.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+	groups.other = unique(groups.other);
+
+	return groups;
+}
+
+function isPublic(id: string) {
+	return [
+		'https://www.w3.org/ns/activitystreams#Public',
+		'as#Public',
```
Comment (mei23) [inline src/remote/activitypub/audience.ts:83]: `as:Public`の間違えだった https://github.com/misskey-dev/misskey/pull/11347

File: src/remote/activitypub/audience.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		'Public',
```
Comment (mei23) [inline src/remote/activitypub/audience.ts:84]: これ (`Public`) いらない疑惑

Comment (mei23) [reply src/remote/activitypub/audience.ts:84]: 元Issueには~~Publicを認識させろとまでは言ってないし~~、
https://www.w3.org/TR/activitypub/#x5-6-public-addressing

いやそんなことなさそうだしうーん
> being represented as simply **Public** or as:Public

Compactしても`as:Public`だし
`Public`が来ても`Canonize` でエラーになるし
```
jsonld.ValidationError: Safe mode validation error.
{
  "type": [
    "JsonLdEvent"
  ],
  "code": "relative object reference",
  "level": "warning",
  "message": "Relative object reference found.",
  "details": {
    "object": "Public"
  }
}
```

[JSON-LD Playground ...](https://json-ld.org/playground/#startTab=tab-canonized&json-ld=%7B%22%40context%22%3A%5B%22https%3A%2F%2Fwww.w3.org%2Fns%2Factivitystreams%22%2C%22https%3A%2F%2Fw3id.org%2Fsecurity%2Fv1%22%2C%7B%22Key%22%3A%22sec%3AKey%22%2C%22manuallyApprovesFollowers%22%3A%22as%3AmanuallyApprovesFollowers%22%2C%22sensitive%22%3A%22as%3Asensitive%22%2C%22Hashtag%22%3A%22as%3AHashtag%22%2C%22toot%22%3A%22http%3A%2F%2Fjoinmastodon.org%2Fns%23%22%2C%22Emoji%22%3A%22toot%3AEmoji%22%2C%22featured%22%3A%22toot%3Afeatured%22%2C%22discoverable%22%3A%22toot%3Adiscoverable%22%2C%22indexable%22%3A%22toot%3Aindexable%22%2C%22schema%22%3A%22http%3A%2F%2Fschema.org%23%22%2C%22PropertyValue%22%3A%22schema%3APropertyValue%22%2C%22value%22%3A%22schema%3Avalue%22%2C%22misskey%22%3A%22https%3A%2F%2Fmisskey-hub.net%2Fns%23%22%2C%22_misskey_content%22%3A%22misskey%3A_misskey_content%22%2C%22_misskey_quote%22%3A%22misskey%3A_misskey_quote%22%2C%22_misskey_reaction%22%3A%22misskey%3A_misskey_reaction%22%2C%22_misskey_votes%22%3A%22misskey%3A_misskey_votes%22%2C%22isCat%22%3A%22misskey%3AisCat%22%2C%22vcard%22%3A%22http%3A%2F%2Fwww.w3.org%2F2006%2Fvcard%2Fns%23%22%2C%22fedibird%22%3A%22http%3A%2F%2Ffedibird.com%2Fns%23%22%2C%22quoteUri%22%3A%22fedibird%3AquoteUri%22%2C%22searchableBy%22%3A%7B%22%40id%22%3A%22fedibird%3AsearchableBy%22%2C%22%40type%22%3A%22%40id%22%7D%7D%5D%2C%22id%22%3A%22https%3A%2F%2Fmisskey.m544.net%2Fnotes%2F718f35151c2b1222f46c0710%22%2C%22url%22%3A%22https%3A%2F%2Fmisskey.m544.net%2Fnotes%2F718f35151c2b1222f46c0710%22%2C%22type%22%3A%22Note%22%2C%22attributedTo%22%3A%22https%3A%2F%2Fmisskey.m544.net%2Fusers%2F5acddf5fcd7aebd97984769a%22%2C%22summary%22%3Anull%2C%22content%22%3A%22%3Cp%3E%E3%81%B5%E3%82%8F%E3%81%81%3C%2Fp%3E%22%2C%22published%22%3A%222024-05-01T16%3A55%3A27.019Z%22%2C%22to%22%3A%5B%22https%3A%2F%2Fmisskey.m544.net%2Fusers%2F5acddf5fcd7aebd97984769a%2Ffollowers%22%5D%2C%22cc%22%3A%5B%22Public%22%5D%2C%22inReplyTo%22%3Anull%2C%22attachment%22%3A%5B%5D%2C%22sensitive%22%3Afalse%2C%22likes%22%3A%22https%3A%2F%2Fmisskey.m544.net%2Fnotes%2F718f35151c2b1222f46c0710%2Flikes%22%2C%22tag%22%3A%5B%5D%7D&frame=%7B%7D&context=%7B%22%40context%22%3A%5B%22https%3A%2F%2Fwww.w3.org%2Fns%2Factivitystreams%22%2C%22https%3A%2F%2Fw3id.org%2Fsecurity%2Fv1%22%2C%7B%22Key%22%3A%22sec%3AKey%22%2C%22manuallyApprovesFollowers%22%3A%22as%3AmanuallyApprovesFollowers%22%2C%22sensitive%22%3A%22as%3Asensitive%22%2C%22Hashtag%22%3A%22as%3AHashtag%22%2C%22toot%22%3A%22http%3A%2F%2Fjoinmastodon.org%2Fns%23%22%2C%22Emoji%22%3A%22toot%3AEmoji%22%2C%22featured%22%3A%22toot%3Afeatured%22%2C%22discoverable%22%3A%22toot%3Adiscoverable%22%2C%22indexable%22%3A%22toot%3Aindexable%22%2C%22schema%22%3A%22http%3A%2F%2Fschema.org%23%22%2C%22PropertyValue%22%3A%22schema%3APropertyValue%22%2C%22value%22%3A%22schema%3Avalue%22%2C%22misskey%22%3A%22https%3A%2F%2Fmisskey-hub.net%2Fns%23%22%2C%22_misskey_content%22%3A%22misskey%3A_misskey_content%22%2C%22_misskey_quote%22%3A%22misskey%3A_misskey_quote%22%2C%22_misskey_reaction%22%3A%22misskey%3A_misskey_reaction%22%2C%22_misskey_votes%22%3A%22misskey%3A_misskey_votes%22%2C%22isCat%22%3A%22misskey%3AisCat%22%2C%22vcard%22%3A%22http%3A%2F%2Fwww.w3.org%2F2006%2Fvcard%2Fns%23%22%2C%22fedibird%22%3A%22http%3A%2F%2Ffedibird.com%2Fns%23%22%2C%22quoteUri%22%3A%22fedibird%3AquoteUri%22%2C%22searchableBy%22%3A%7B%22%40id%22%3A%22fedibird%3AsearchableBy%22%2C%22%40type%22%3A%22%40id%22%7D%7D%5D%7D)

**Pull request merged**